### DISCUSSION TITLE
How to render links of IPv6 Addresses?

### DISCUSSION THREAD

**Discussion by produnis** [quarto-dev/quarto-cli#9229] [Q&A] [answered]
### Description

Hi there, I have IPv6-Addresses in my document like this:

```

:::{.linkelement}
[![](img/qnap.jpg "NAS")](http://[fd04::123]:8080)
:::

```

If I render the document, the link is generated as:
`http://%5Bfd04::123%5D:8080`

How would I link IPv6 addresses correctly?

**Comment by mcanouil** [comment]
For what is related to Quarto/Pandoc, the link syntaxes are:
- `[Text to display](ULR)`
- `<URL>`
- `` `<a href="URL">Text to display</a>`{=html}``

<details><summary><i>Outdated</i></summary>
<p>

> If I render the document, the link is generated as:
> http://%5Bfd04::123%5D:8080

To which format? It's really important to provide complete example as here the format is not specified anywhere.

</p>
</details>

**Reply by mcanouil** [reply]
Here is a reproducible example that could be use to report the issue in [Pandoc repository](https://github.com/jgm/pandoc):

```bash
echo "<http://[fd04::123]:8080>" > index.md
pandoc index.md --from markdown --to html
```

```html
<p><a href="http://%5Bfd04::123%5D:8080" class="uri">http://[fd04::123]:8080</a></p>
```

Using raw code works as Pandoc is not parsing it:

```bash
echo '`<a href="http://[fd04::123]:8080">http://[fd04::123]:8080</a>`{=html}' > index.md
pandoc index.md --from markdown --to html
```

```html
<p><a href="http://[fd04::123]:8080">http://[fd04::123]:8080</a></p>
```

**Reply by cderv** [reply, marked as answer]
Definely something with Pandoc which HTML escape the characters when parsing the link 
````
echo '[test](http://[fd04::123]:8080)' | pandoc -f markdown -t native                            
[ Para
    [ Link
        ( "" , [] , [] )
        [ Str "test" ]
        ( "http://%5Bfd04::123%5D:8080" , "" )
    ]
]
````

They do escape at reading time with usual URI escaping 
https://github.com/jgm/pandoc/blob/e4ac26d255ad8748c01301b3319975de39a8f43f/src/Text/Pandoc/URI.hs#L29-L32

Note that JS function also does it 
````
> encodeURI('http://[fd04::123]:8080')
'http://%5Bfd04::123%5D:8080'
````

So I think those URL are quite specific, right ? 

It seems it requires a specific handling for ipv6 following recent [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986)
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI#encoding_for_rfc3986

So it definitely need an update to Pandoc to follow ipv6 compliance if possible. 

but you could also for now do that yourself using Lua filter

````lua
-- ipv6.lua
Link = function(l)
    if l.classes:includes("ipv6") then
    	_, i = l.classes:find("ipv6")
        l.classes:remove(i)
        l.target = l.target:gsub("%%5B", "[")
        l.target = l.target:gsub("%%5D", "]")
    end
    return l
end 
````

````powershell
❯ echo '[test](http://\[fd04::123\]:8080){.ipv6}' | pandoc -f markdown -t native --lua-filter=ipv6.lua
[ Para
    [ Link
        ( "" , [] , [] )
        [ Str "test" ]
        ( "http://[fd04::123]:8080" , "" )
    ]
]
````

````powershell
❯ echo '[test](http://\[fd04::123\]:8080){.ipv6}' | pandoc -f markdown -t html --lua-filter=ipv6.lua  
<p><a href="http://[fd04::123]:8080">test</a></p>
````

Hope it helps

**Reply by produnis** [reply]
Thx a lot for that lua snippet!!!! Works great for me!

**Reply by cscheid** [reply]
That `ipv6` class trick is really nice. I wonder if we should just have that as a Quarto feature...

**Reply by mcanouil** [reply]
Having this as a Quarto feature rather than fixing it upstream? 🤔
